File: src/main/java/tooko/main/extras/SwitchLang.java
package tooko.main.extras;

import tooko.main.*;
import tooko.td.TdApi.*;
import tooko.td.client.*;
import tooko.td.core.*;

public class SwitchLang extends TdHandler {

    public static int DATA_ID = Fn.DataId._1;
    public static InlineArray langs;

    @Override
    public void onLoad() {

        initFunction("lang");

        initData(DATA_ID);

        if (langs == null) {

            langs = new InlineArray();

            for (Lang L : Lang.ALL.values()) {

                langs.dataLine(L.LANG_NAME, 0, L.LANG_ID);

            }

        }

    }

    @Override
    public void onFunction(User user, long chatId, Message message, String function, String param, String[] params, String[] originParams) {

        send(Fn.sendText(chatId, SwitchLang.langs, Fn.plainText(Lang.get(user).CHOOSE_LANG)));
    }

    @Override
    public void onNewCallbackQuery(int userId, long chatId, long messageId, long queryId, int subId, byte[][] data) {

        send(Fn.answerConfirm(queryId));

        Lang.DATA.setById(userId, new Lang.DB(userId, subId));

        Lang L = Lang.get(userId);

        FormattedText msg = Fn.parseHtml(L.LANG_SELECTED);

        try {

            execute(Fn.editText(chatId, messageId, msg));

            send(Fn.sendText(chatId, Fn.plainText(L.LAUNCH)));

        } catch (TdException ex) {

            if (ex.getCode() == -1) {

                send(Fn.deleteMessages(chatId, messageId));

                send(Fn.answerServerClosed(queryId, userId));

            } else {

                send(Fn.sendText(chatId, msg));

                send(Fn.deleteMessages(chatId, messageId));

            }

        }

    }

}

File: src/main/java/tooko/core/funs/SwitchLang.kt
/*
 * Copyright (c) 2019 - 2020 KazamaWataru
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tooko.core.funs

import td.TdApi
import tooko.core.DATA_4
import tooko.core.client.TdBotHandler
import tooko.core.env.Lang
import tooko.core.fromChannel
import tooko.core.fromPrivate
import tooko.core.langFor
import tooko.core.utils.inlineButton
import tooko.core.utils.make

class SwitchLang : TdBotHandler() {

    val DATA_ID = DATA_4

    override fun onLoad() {

        initFunction("lang")

        initData(DATA_ID)

    }

    override suspend fun onFunction(userId: Int, chatId: Long, message: TdApi.Message, function: String, param: String, params: Array<String>, originParams: Array<String>) {

        val L = chatId.langFor

        if (message.fromPrivate) {

            sudo make L.CHOOSE_LANG withMarkup inlineButton {

                Lang.BY_NAME.forEach { (name, lang) ->

                    dataLine(name, DATA_ID, lang.LANG_ID)

                }

            } sendTo chatId

        } else if (message.fromChannel) {

            sudo make L.CHOOSE_CHANNEL_LANG withMarkup inlineButton {

                Lang.BY_NAME.forEach { (name, lang) ->

                    dataLine(name, DATA_ID, lang.LANG_ID)

                }

            } sendTo chatId

        } else {

            sudo make L.CHOOSE_GROUP_LANG withMarkup inlineButton {

                Lang.BY_NAME.forEach { (name, lang) ->

                    dataLine(name, DATA_ID, lang.LANG_ID)

                }

            } sendTo chatId


        }

    }

    override suspend fun onNewCallbackQuery(userId: Int, chatId: Long, messageId: Long, queryId: Long, subId: Int, data: Array<ByteArray>) {

        val selected = subId.langFor

        Lang.DATA.setById(chatId, Lang.DB(chatId, subId))

        sudo make selected.LANG_SELECTED at messageId editTo chatId

    }

}